modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientPartitionAwarenessDiscoveryTest.java

```diff
@@ -0,0 +1,141 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.client.thin;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.ignite.client.ClientAddressFinder;
+import org.apache.ignite.client.ClientCache;
+import org.apache.ignite.configuration.ClientConfiguration;
+import org.apache.ignite.internal.IgnitionEx;
+import org.apache.ignite.internal.processors.cache.IgniteInternalCache;
+import org.junit.Test;
+
+import static org.apache.ignite.internal.processors.odbc.ClientListenerProcessor.CLIENT_LISTENER_PORT;
+
+/**
+ * Test partition awareness of thin client on changed topology.
+ */
+public class ThinClientPartitionAwarenessDiscoveryTest extends ThinClientAbstractPartitionAwarenessTest {
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /**
+     * Test that client use channels to all running nodes while new nodes start
+     */
+    @Test
+    public void testClientDiscoveryNodesJoin() throws Exception {
+        for (int i = 0; i < MAX_CLUSTER_SIZE; ++i) {
+            startGrid(i);
+            awaitPartitionMapExchange();
+
+            int[] workChannels = IntStream.rangeClosed(0, i).toArray();
+
+            if (i == 0)
+                initClient(getClientConfigurationWithDiscovery(), workChannels);
+            else {
+                detectTopologyChange();
+                awaitChannelsInit(workChannels);
+            }
+
+            testPartitionAwareness(workChannels);
+        }
+    }
+
+    /**
+     * Test that client use channels to all running nodes while nodes stop
+     */
+    @Test
+    public void testClientDiscoveryNodesLeave() throws Exception {
+        startGrids(MAX_CLUSTER_SIZE);
+        awaitPartitionMapExchange();
+
+        initClient(getClientConfigurationWithDiscovery(), 0, 1, 2, 3);
+        detectTopologyChange();
+
+        for (int i = MAX_CLUSTER_SIZE - 1; i != 0; i--) {
+            int[] workChannels = IntStream.range(0, i).toArray();
+
+            channels[i] = null;
+            stopGrid(i);
+            awaitPartitionMapExchange();
+            detectTopologyChange();
+
+            awaitChannelsInit(workChannels);
```

Comment:
All channels already inited after initClient, what else do we wait here?

Reply:
fixied